middle_end/flambda2/kinds/flambda_kind.ml

```diff
@@ -173,102 +163,173 @@ let is_naked_float t =
   | Region | Rec_info ->
     false
 
+module Flat_suffix_element = struct
+  type kind = t
+
+  type t =
+    | Tagged_immediate
+    | Naked_float
+    | Naked_float32
+    | Naked_int32
+    | Naked_int64
+    | Naked_nativeint
+
+  let kind t =
+    match t with
+    | Tagged_immediate -> value
+    | Naked_float -> naked_float
+    | Naked_float32 -> naked_float32
+    | Naked_int32 -> naked_int32
+    | Naked_int64 -> naked_int64
+    | Naked_nativeint -> naked_nativeint
+
+  let compare = Stdlib.compare
+
+  let equal = Stdlib.( = )
+
+  let print ppf t =
+    match t with
+    | Tagged_immediate -> Format.pp_print_string ppf "Tagged_immediate"
+    | Naked_float -> Format.pp_print_string ppf "Naked_float"
+    | Naked_float32 -> Format.pp_print_string ppf "Naked_float32"
+    | Naked_int32 -> Format.pp_print_string ppf "Naked_int32"
+    | Naked_int64 -> Format.pp_print_string ppf "Naked_int64"
+    | Naked_nativeint -> Format.pp_print_string ppf "Naked_nativeint"
+
+  let from_lambda (elt : Lambda.flat_element) =
+    match elt with
+    | Imm -> Tagged_immediate
+    | Float_boxed | Float64 -> Naked_float
+    | Float32 -> Naked_float32
+    | Bits32 -> Naked_int32
+    | Bits64 -> Naked_int64
+    | Word -> Naked_nativeint
+end
+
 module Mixed_block_shape = struct
   type kind = t
 
   type t =
-    { fields : kind array;
-      (* For compiling to Cmm, we need the lambda shape. We also use it to know
-         the value prefix length. *)
-      lambda_shape : Lambda.mixed_block_shape
+    { value_prefix_size : int;
+      flat_suffix : Flat_suffix_element.t array;
+      field_kinds : kind array
     }
 
-  let field_kinds { fields; lambda_shape = _ } = fields
+  let value_prefix_size t = t.value_prefix_size
 
-  let value_prefix_size { fields = _; lambda_shape } =
-    lambda_shape.value_prefix_len
+  let flat_suffix t = t.flat_suffix
 
-  let equal t1 t2 =
-    Int.equal t1.lambda_shape.value_prefix_len t2.lambda_shape.value_prefix_len
-    && Int.equal (Array.length t1.fields) (Array.length t2.fields)
-    && Array.for_all2 equal t1.fields t2.fields
+  let field_kinds t = t.field_kinds
 
-  let compare t1 t2 =
-    let c =
-      Int.compare t1.lambda_shape.value_prefix_len
-        t2.lambda_shape.value_prefix_len
-    in
+  let equal
+      { value_prefix_size = value_prefix_size1;
+        flat_suffix = flat_suffix1;
+        field_kinds = _
+      }
+      { value_prefix_size = value_prefix_size2;
+        flat_suffix = flat_suffix2;
+        field_kinds = _
+      } =
+    (* [field_kinds] is uniquely determined by [flat_suffix]. *)
```

Comment:
Could you move this comment to the type definition ? Maybe elaborating a bit more, saying that `field_kinds` is what gets accessed most during `Flambda` but the other two are necessary to compile to Cmm.

Reply:
Done